torch/nn/parallel/distributed.py

```diff
@@ -442,10 +442,24 @@ def __init__(self, module, device_ids=None,
         # Whether to perform input tensor CPU to GPU copies on a side-stream
         self.use_side_stream_for_tensor_copies = os.environ.get("PYTORCH_DDP_USE_SIDE_STREAM", "1") == "1"
 
-        # Sync params and buffers
+        # Module replication within process (single-process multi device)
+        self._module_copies = self._replicate_modules_within_process()
+        # Build parameters for reducer.
+        parameters, expect_sparse_gradient = self._build_params_for_reducer()
+        # Verify model equivalence.
+        # Corresponding params' layouts (strides) must match across
+        # replicas within this process and across processes.
+        # see Note: "Gradient Layout Contract" in Reducer::initialize_buckets).
+        if self.device_ids is not None and len(self.device_ids) > 1:
+            dist._verify_replicas_within_process(parameters, expect_sparse_gradient)
+        self._verify_model_across_ranks(parameters)
```

Comment:
nit: directly call dist._verify_replica0_across_processes(self.process_group, parameters)?

Reply:
The motivation for helper function was that after this diff, we will add CPU based barrier (https://github.com/pytorch/pytorch/issues/52876)  in debug mode, to report which ranks had an inconsistent model to help with a debug (and for all processes to throw a deterministic error that can be handled)

Reply:
Changed it to directly call the cpp function for now, for the above comment, we can address it in that diff itself.